annotate.el

```diff
@@ -1073,6 +1090,116 @@ example:
         (when annotate-use-messages
           (message "Annotations loaded."))))))
 
+(defun annotate-db-clean-records (records-db)
+  "Remove records from arg `records-db' that have empty annotation, example:
+
+'((\"/foo/bar.dat\" nil \"abababababababababababababab\")
+  (\"/foo/baz.dat\" ((0 9 \"note\" \"annotated\")) \"abababababababababababababab\"))
+
+will become:
+
+'((\"/foo/baz.dat\" ((0 9 \"note\" \"annotated\")) \"abababababababababababababab\"))
+
+i.e. the first record is removed."
+  (cl-remove-if (lambda (a) (null (annotate-annotations-from-dump a)))
+                records-db))
+
+(defun annotate-db-purge ()
+ "Update datbase *on disk* removing all the records with empty
+annotation."
+  (interactive)
+  (let ((db (annotate-db-clean-records (annotate-load-annotation-data))))
+    (annotate-dump-annotation-data db)))
+
+(defun annotate-load-annotation-data ()
+  "Read and return saved annotations."
+  (with-temp-buffer
+    (when (file-exists-p annotate-file)
+      (insert-file-contents annotate-file))
+    (goto-char (point-max))
+    (cond ((= (point) 1)
+           nil)
+          (t
+           (goto-char (point-min))
+           (read (current-buffer))))))
+
+(defun annotate-dump-annotation-data (data)
+  "Save `data` into annotation file."
+  (with-temp-file annotate-file
+    (let ((print-length nil))
+      (prin1 data (current-buffer)))))
+
+(defun annotate-db-remove-annotation (db-records
+                                      record-filename
+                                      annotation-beginning
+                                      annotation-ending)
+  "Remove from database `db-records' the annotation indentified by
+ the triplets `record-filename', `annotation-beginning' and
```

Comment:
What happens if the triplet does not match? Can this happen?

Reply:
Hi!
Yes this is likely to happens!

If `file-matched-record` is nil the value of the function is just `db-records` (no changes, so far so good)

if `file-matched-record` is not nil (i.e. its value is the record that contains `record-filename`) and
there is no annotation in said record with limits equals to `annotation-beginning` and 
`annotation-ending` then:
- `rest-of-db` values the db without the record that match `record-filename` (and this record must exists as `file-matched-record` is non-nil);
- new-annotations contains exactly the same annotation as in the original record (because `cl-remove-if` is going to remove nothing).

so here:

``` lisp
(new-record  (annotate-make-record record-filename
                                   new-annotations
                                   checksum)))
```
`new-record` is equals to the original record with `record-filename` and

```
(push new-record
      rest-of-db))
```
the value of this form is equivalent to `db-records`, nothing changed actually.
If that there is no fallacy in that logic (nor really sure there was not :-) ) if there is no match the database returned is always at least equivalent to the one passed as argument.

Do you agree? Because i am not totally sure now that i wrote that! :-D :-D

Bye!
C.

Reply:
Seems good, thank you for clarifying.

It *feels* like there's a cleaner way to express this. All the `find-if` and `remove-if` make my head spin a little. But maybe that's also just a consequence of the `cl-` functions available to us. Anyway, this looks good to me.